lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,783 @@
+#include "seahorn/PredicateAbstraction.hh"
+#include "seahorn/HornifyModule.hh"
+#include "seahorn/HornClauseDBTransf.hh"
+#include "seahorn/HornClauseDB.hh"
+
+#include "llvm/IR/Function.h"
+#include "llvm/Support/CommandLine.h"
+
+#include "ufo/Expr.hpp"
+#include "ufo/Smt/Z3n.hpp"
+#include "ufo/Smt/EZ3.hh"
+#include <vector>
+#include <boost/logic/tribool.hpp>
+#include "seahorn/HornClauseDBWto.hh"
+#include <algorithm>
+
+#include "ufo/Stats.hh"
+
+using namespace llvm;
+
+namespace seahorn
+{
+	char PredicateAbstraction::ID = 0;
+
+	bool PredicateAbstraction::runOnModule (Module &M)
+	{
+		HornifyModule &hm = getAnalysis<HornifyModule> ();
+
+		PredicateAbstractionAnalysis pabs(hm);
+		pabs.runAnalysis();
+
+		return false;
+	}
+
+	void PredicateAbstraction::getAnalysisUsage (AnalysisUsage &AU) const
+	{
+		AU.addRequired<HornifyModule> ();
+		AU.setPreservesAll();
+	}
+
+	void PredicateAbstractionAnalysis::runAnalysis()
+	{
+		Stats::resume ("Pabs solve");
+		//load the Horn clause database
+		auto &db = m_hm.getHornClauseDB ();
+		db.buildIndexes ();
+
+		//guess candidates
+		guessCandidate(db);
+
+		HornDbModel oldModel;
+		oldModel.setCurrentCandidates(currentCandidates);
+
+		PredAbsHornModelConverter converter;
+
+		//run main algorithm
+		HornClauseDB new_db = generateAbstractDB(db, converter);
+
+		//initialize spacer based on new DB
+		m_fp.reset (new ZFixedPoint<EZ3> (m_hm.getZContext ()));
```

Comment:
I see how `fp` is initialized. This is enough for a quick experiment, but not sustainable in the long run.